src/Microsoft.OData.Core/Json/JsonLightInstanceAnnotationWriter.cs

```diff
@@ -92,19 +92,99 @@ internal JsonLightInstanceAnnotationWriter(ODataJsonLightValueSerializer valueSe
             Debug.Assert(instanceAnnotations != null, "instanceAnnotations should not be null if we called this");
             Debug.Assert(tracker != null, "tracker should not be null if we called this");
 
-            HashSet<string> instanceAnnotationNames = new HashSet<string>(StringComparer.Ordinal);
-            foreach (var annotation in instanceAnnotations)
+            WriteInstanceAnnotationsInternal(instanceAnnotations, tracker, ignoreFilter, propertyName);
+        }
+
+        /// <summary>
+        /// Writes all the instance annotations specified in <paramref name="instanceAnnotations"/>
+        /// </summary>
+        /// <param name="instanceAnnotations">Collection of instance annotations to write</param>
+        /// <param name="tracker">The tracker that tracks if instance annotations are written</param>
+        /// <param name="instanceAnnotationNames">The set of nam</param>
+        /// <param name="ignoreFilter"></param>
+        /// <param name="propertyName"></param>
+        private void WriteInstanceAnnotationsInternal(
+            IEnumerable<ODataInstanceAnnotation> instanceAnnotations,
+            InstanceAnnotationWriteTracker tracker,
+            bool ignoreFilter = false,
+            string propertyName = null)
+        {
+            // this method runs in a hot path hence the optimizations
+
+            // initialize to null to avoid instatiating a HashSet if there are no annotations
+            HashSet<string> instanceAnnotationNames = null;
+            
+            // this method is called with a List most of the time
+            // foreach against a List does not allocate the enumerator to the heap,
+            // but foreach against an IEnumerable does
+            if (instanceAnnotations is List<ODataInstanceAnnotation> instanceAnnotationsList)
             {
-                if (!instanceAnnotationNames.Add(annotation.Name))
+                foreach (var annotation in instanceAnnotationsList)
                 {
-                    throw new ODataException(ODataErrorStrings.JsonLightInstanceAnnotationWriter_DuplicateAnnotationNameInCollection(annotation.Name));
+                    if (instanceAnnotationNames == null)

```

Comment:
Let explore this option.

Reply:
Could you clarify what you meant by the last part "This should also eliminate the `if` evaluations for all iterations". Which `if` statements specifically do you suggest will be eliminated? The `if (instanceAnnotationNames == null)` used to conditionally instantiate the hash set, or the `if (instanceAnnotations is List<...>)`?

Reply:
I used `ICollection` as the argument type and used `if (instanceAnnotations?.Count > 0)` to conditionally create the HashSet. However, the I still retain `if (instanceAnnotations is List<...>)` because a `foreach` against an `ICollection` still allocates an enumerator, and `ICollection` does not expose an indexer so we can't use a `for (int i...)`